lib/charms/tls_certificates_interface/v4/tls_certificates.py

```diff
@@ -357,6 +345,90 @@ def from_string(private_key: str) -> "PrivateKey":
         return PrivateKey(raw=private_key.strip())
 
 
+@dataclass(frozen=True)
+class Certificate:
+    """This class represents a certificate."""
+
+    raw: str
+    common_name: str
+    sans_dns: Optional[List[str]] = None
+    sans_ip: Optional[List[str]] = None
+    sans_oid: Optional[List[str]] = None
+    email_address: Optional[str] = None
+    organization: Optional[str] = None
+    organizational_unit: Optional[str] = None
+    country_name: Optional[str] = None
+    state_or_province_name: Optional[str] = None
+    locality_name: Optional[str] = None
+    expiry_time: Optional[datetime] = None
+    validity_start_time: Optional[datetime] = None
+
+    def __str__(self) -> str:
+        """Return the certificate as a string."""
+        return self.raw
+
+    @staticmethod
+    def from_string(certificate: str) -> "Certificate":
```

Comment:
```suggestion
    @classmethod
    def from_string(cls: "Certificate", certificate: str) -> "Certificate":
```

This would be a bit more idiomatic, the method is really on the class itself as it is an alternative constructor.

Reply:
I'm not sure if I'm getting something wrong but with the approach you propose, I'm getting these static analysis errors:

```
/home/guillaume/code/tls-certificates-interface/tests/unit/charms/tls_certificates_interface/v4/test_tls_certificates_v4_requires.py
  /home/guillaume/code/tls-certificates-interface/tests/unit/charms/tls_certificates_interface/v4/test_tls_certificates_v4_requires.py:205:70 - error: Cannot access attribute "from_string" for class "type[Certificate]"
    Could not bind method "from_string" because "type[Certificate]" is not assignable to parameter "cls"
      "type[type]" is incompatible with "type[Certificate]" (reportAttributeAccessIssue)
  /home/guillaume/code/tls-certificates-interface/tests/unit/charms/tls_certificates_interface/v4/test_tls_certificates_v4_requires.py:206:61 - error: Cannot access attribute "from_string" for class "type[Certificate]"
    Could not bind method "from_string" because "type[Certificate]" is not assignable to parameter "cls"
      "type[type]" is incompatible with "type[Certificate]" (reportAttributeAccessIssue)
/home/guillaume/code/tls-certificates-interface/tests/unit/charms/tls_certificates_interface/v4/dummy_provider_charm/src/charm.py
  /home/guillaume/code/tls-certificates-interface/tests/unit/charms/tls_certificates_interface/v4/dummy_provider_charm/src/charm.py:116:41 - error: Cannot access attribute "from_string" for class "type[Certificate]"
    Could not bind method "from_string" because "type[Certificate]" is not assignable to parameter "cls"
      "type[type]" is incompatible with "type[Certificate]" (reportAttributeAccessIssue)
  /home/guillaume/code/tls-certificates-interface/tests/unit/charms/tls_certificates_interface/v4/dummy_provider_charm/src/charm.py:118:32 - error: Cannot access attribute "from_string" for class "type[Certificate]"
    Could not bind method "from_string" because "type[Certificate]" is not assignable to parameter "cls"
      "type[type]" is incompatible with "type[Certificate]" (reportAttributeAccessIssue)
  /home/guillaume/code/tls-certificates-interface/tests/unit/charms/tls_certificates_interface/v4/dummy_provider_charm/src/charm.py:120:33 - error: Cannot access attribute "from_string" for class "type[Certificate]"
    Could not bind method "from_string" because "type[Certificate]" is not assignable to parameter "cls"
      "type[type]" is incompatible with "type[Certificate]" (reportAttributeAccessIssue)
/home/guillaume/code/tls-certificates-interface/tests/integration/v4/provider_charm/src/charm.py
  /home/guillaume/code/tls-certificates-interface/tests/integration/v4/provider_charm/src/charm.py:77:49 - error: Cannot access attribute "from_string" for class "type[Certificate]"
    Could not bind method "from_string" because "type[Certificate]" is not assignable to parameter "cls"
      "type[type]" is incompatible with "type[Certificate]" (reportAttributeAccessIssue)
  /home/guillaume/code/tls-certificates-interface/tests/integration/v4/provider_charm/src/charm.py:79:40 - error: Cannot access attribute "from_string" for class "type[Certificate]"
    Could not bind method "from_string" because "type[Certificate]" is not assignable to parameter "cls"
      "type[type]" is incompatible with "type[Certificate]" (reportAttributeAccessIssue)
  /home/guillaume/code/tls-certificates-interface/tests/integration/v4/provider_charm/src/charm.py:81:41 - error: Cannot access attribute "from_string" for class "type[Certificate]"
    Could not bind method "from_string" because "type[Certificate]" is not assignable to parameter "cls"
      "type[type]" is incompatible with "type[Certificate]" (reportAttributeAccessIssue)
  /home/guillaume/code/tls-certificates-interface/tests/integration/v4/provider_charm/src/charm.py:82:41 - error: Cannot access attribute "from_string" for class "type[Certificate]"
    Could not bind method "from_string" because "type[Certificate]" is not assignable to parameter "cls"
      "type[type]" is incompatible with "type[Certificate]" (reportAttributeAccessIssue)
/home/guillaume/code/tls-certificates-interface/lib/charms/tls_certificates_interface/v4/tls_certificates.py
  /home/guillaume/code/tls-certificates-interface/lib/charms/tls_certificates_interface/v4/tls_certificates.py:284:37 - error: Cannot access attribute "from_string" for class "type[Certificate]"
    Could not bind method "from_string" because "type[Certificate]" is not assignable to parameter "cls"
      "type[type]" is incompatible with "type[Certificate]" (reportAttributeAccessIssue)
  /home/guillaume/code/tls-certificates-interface/lib/charms/tls_certificates_interface/v4/tls_certificates.py:288:28 - error: Cannot access attribute "from_string" for class "type[Certificate]"
    Could not bind method "from_string" because "type[Certificate]" is not assignable to parameter "cls"
      "type[type]" is incompatible with "type[Certificate]" (reportAttributeAccessIssue)
  /home/guillaume/code/tls-certificates-interface/lib/charms/tls_certificates_interface/v4/tls_certificates.py:289:32 - error: Cannot access attribute "from_string" for class "type[Certificate]"
    Could not bind method "from_string" because "type[Certificate]" is not assignable to parameter "cls"
      "type[type]" is incompatible with "type[Certificate]" (reportAttributeAccessIssue)
  /home/guillaume/code/tls-certificates-interface/lib/charms/tls_certificates_interface/v4/tls_certificates.py:370:6 - error: Argument of type "(cls: Certificate, certificate: str) -> Certificate" cannot be assigned to parameter "f" of type "(type[_T@classmethod], **_P@classmethod) -> _R_co@classmethod" in function "__init__"
    Type "(cls: Certificate, certificate: str) -> Certificate" is incompatible with type "(type[_T@classmethod], **_P@classmethod) -> _R_co@classmethod"
      Parameter 1: type "type[_T@classmethod]" is incompatible with type "Certificate"
        "type[type]" is incompatible with "type[Certificate]" (reportArgumentType)
  /home/guillaume/code/tls-certificates-interface/lib/charms/tls_certificates_interface/v4/tls_certificates.py:371:26 - error: Type of parameter "cls" must be a supertype of its class "type[Certificate]" (reportGeneralTypeIssues)
  /home/guillaume/code/tls-certificates-interface/lib/charms/tls_certificates_interface/v4/tls_certificates.py:414:16 - error: Object of type "Certificate" is not callable
    Attribute "__call__" is unknown (reportCallIssue)
  /home/guillaume/code/tls-certificates-interface/lib/charms/tls_certificates_interface/v4/tls_certificates.py:735:40 - error: Cannot access attribute "from_string" for class "type[Certificate]"
    Could not bind method "from_string" because "type[Certificate]" is not assignable to parameter "cls"
      "type[type]" is incompatible with "type[Certificate]" (reportAttributeAccessIssue)
  /home/guillaume/code/tls-certificates-interface/lib/charms/tls_certificates_interface/v4/tls_certificates.py:739:31 - error: Cannot access attribute "from_string" for class "type[Certificate]"
    Could not bind method "from_string" because "type[Certificate]" is not assignable to parameter "cls"
      "type[type]" is incompatible with "type[Certificate]" (reportAttributeAccessIssue)
  /home/guillaume/code/tls-certificates-interface/lib/charms/tls_certificates_interface/v4/tls_certificates.py:741:35 - error: Cannot access attribute "from_string" for class "type[Certificate]"
    Could not bind method "from_string" because "type[Certificate]" is not assignable to parameter "cls"
      "type[type]" is incompatible with "type[Certificate]" (reportAttributeAccessIssue)
18 errors, 0 warnings, 0 informations 
```

```python

    @classmethod
    def from_string(cls: "Certificate", certificate: str) -> "Certificate":
        """Create a Certificate object from a certificate."""
        try:
            certificate_object = x509.load_pem_x509_certificate(data=certificate.encode())
        except ValueError as e:
            logger.error("Could not load certificate: %s", e)
            raise TLSCertificatesError("Could not load certificate")
        common_name = certificate_object.subject.get_attributes_for_oid(NameOID.COMMON_NAME)
        country_name = certificate_object.subject.get_attributes_for_oid(NameOID.COUNTRY_NAME)
        state_or_province_name = certificate_object.subject.get_attributes_for_oid(
            NameOID.STATE_OR_PROVINCE_NAME
        )
        locality_name = certificate_object.subject.get_attributes_for_oid(NameOID.LOCALITY_NAME)
        organization_name = certificate_object.subject.get_attributes_for_oid(
            NameOID.ORGANIZATION_NAME
        )
        email_address = certificate_object.subject.get_attributes_for_oid(NameOID.EMAIL_ADDRESS)
        try:
            sans = certificate_object.extensions.get_extension_for_class(
                x509.SubjectAlternativeName
            ).value
            sans_dns = [
                str(san)
                for san in sans.get_values_for_type(x509.DNSName)
                if isinstance(san, x509.DNSName)
            ]
            sans_ip = [
                str(san)
                for san in sans.get_values_for_type(x509.IPAddress)
                if isinstance(san, x509.IPAddress)
            ]
            sans_oid = [
                str(san)
                for san in sans.get_values_for_type(x509.RegisteredID)
                if isinstance(san, x509.RegisteredID)
            ]
        except x509.ExtensionNotFound:
            sans_dns = []
            sans_ip = []
            sans_oid = []
        expiry_time = certificate_object.not_valid_after_utc
        validity_start_time = certificate_object.not_valid_before_utc

        return cls(
            raw=certificate.strip(),
            common_name=str(common_name[0].value),
            country_name=str(country_name[0].value) if country_name else None,
            state_or_province_name=str(state_or_province_name[0].value)
            if state_or_province_name
            else None,
            locality_name=str(locality_name[0].value) if locality_name else None,
            organization=str(organization_name[0].value) if organization_name else None,
            email_address=str(email_address[0].value) if email_address else None,
            sans_dns=sans_dns,
            sans_ip=sans_ip if sans_ip else None,
            sans_oid=sans_oid if sans_oid else None,
            expiry_time=expiry_time,
            validity_start_time=validity_start_time,
        )

```

Reply:
My bad, the simple way to avoid those typing issues was not to type the `cls` attribute. 

For example:
```python

    @classmethod
    def from_string(cls, private_key: str) -> "PrivateKey":
        """Create a PrivateKey object from a private key."""
        return cls(raw=private_key.strip())
```

Reply:
This is done